File: cli/src/ffprobe.ts
import { spawn } from 'node:child_process';

/**
 * Subset of metadata extracted via ffprobe. Used for the sidecar
 * `videos/<filename>.meta.json` and to drive thumbnail generation
 * (we need duration to pick a sensible seek time for short clips).
 */
export interface ProbeResult {
  duration: number | null;
  width: number | null;
  height: number | null;
}

/**
 * Run `ffprobe` against a file and return basic stream/format info.
 * Throws if ffprobe is not on PATH or exits non-zero.
 */
export async function probeVideo(filePath: string): Promise<ProbeResult> {
  return new Promise((resolveResult, reject) => {
    const proc = spawn('ffprobe', [
      '-v',
      'error',
      '-select_streams',
      'v:0',
      '-show_entries',
      'stream=width,height',
      '-show_entries',
      'format=duration',
      '-of',
      'json',
      filePath,
    ]);

    let stdout = '';
    let stderr = '';
    proc.stdout.on('data', (d) => (stdout += d.toString()));
    proc.stderr.on('data', (d) => (stderr += d.toString()));

    proc.on('error', (err) => {
      reject(new Error(`ffprobe spawn failed: ${err.message}. Is ffprobe installed?`));
    });

    proc.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`ffprobe exited ${code}: ${stderr.trim()}`));
        return;
      }
      try {
        const data = JSON.parse(stdout);
        const stream = data.streams?.[0] ?? {};
        const format = data.format ?? {};
        resolveResult({
          duration: format.duration ? parseFloat(format.duration) : null,
          width: stream.width ?? null,
          height: stream.height ?? null,
        });
      } catch (err) {
        reject(new Error(`failed to parse ffprobe output: ${(err as Error).message}`));
      }
    });
  });
}

File: cli-go/internal/ffprobe/ffprobe.go
package ffprobe

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
)

// ProbeResult holds basic stream/format info extracted via ffprobe.
type ProbeResult struct {
	Duration *float64 `json:"duration"`
	Width    *int     `json:"width"`
	Height   *int     `json:"height"`
}

// internal ffprobe JSON structures.
type ffprobeOutput struct {
	Streams []struct {
		Width  *int `json:"width"`
		Height *int `json:"height"`
	} `json:"streams"`
	Format struct {
		Duration string `json:"duration"`
	} `json:"format"`
}

// ProbeVideo runs ffprobe against a file and returns duration/dimensions.
func ProbeVideo(filePath string) (*ProbeResult, error) {
	cmd := exec.Command("ffprobe",
		"-v", "error",
		"-select_streams", "v:0",
		"-show_entries", "stream=width,height",
		"-show_entries", "format=duration",
		"-of", "json",
		filePath,
	)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		if _, ok := err.(*exec.Error); ok {
			return nil, fmt.Errorf("ffprobe spawn failed: %w. Is ffprobe installed?", err)
		}
		msg := stderr.String()
		if msg == "" {
			msg = err.Error()
		}
		return nil, fmt.Errorf("ffprobe exited: %s", msg)
	}

	var data ffprobeOutput
	if err := json.Unmarshal(stdout.Bytes(), &data); err != nil {
		return nil, fmt.Errorf("failed to parse ffprobe output: %w", err)
	}

	result := &ProbeResult{}
	if len(data.Streams) > 0 {
		result.Width = data.Streams[0].Width
		result.Height = data.Streams[0].Height
	}
	if data.Format.Duration != "" {
		if d, err := strconv.ParseFloat(data.Format.Duration, 64); err == nil {
			result.Duration = &d
		}
	}

	return result, nil
}
